HPCC-17625 JPTree memory usage and contention issues

richardkchapman opened the pull request:
Split into several commits for ease of review. Reduces memory consumption by around 75% on my simple tests, and also reduces contention which seems to give a reasonable performance lift.

I'm marking this as a bugfix as memory usage is unacceptable in current LocalPTree usage.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [x] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
Unit tests, full regression runs, valgrind.
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Allow PTree priority (speed vs memory) to be selected at creation time.

Create some baseline timings to allow further investigation of contention and
evaluation of potential improvements.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Add tracing to help check memory usage.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Avoid storing (unused) hash and linkcount in attribute keys and values for
non-atomtable PTrees.

Reduces memory usage by approximately 50% on my testing...

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Avoid storing (unused) hash and linkcount in element names for
non-atomtable PTrees.

Reduces memory usage for names by approximately 50% on my testing...

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Store short strings in place of the pointer, for names, attribute keys, and
values.

Reduces memory usage by another 50% or so on my tests...

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Code from prior commit cleaned up and turned into a template.

Common up the stats for simplicity.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Experiment with moving from a single critical block to smaller ones. Should
reduce contention, but may also cause us to spend more time
acquiring/releaseing locks.

It seems that smaller locking sections made things slower, but removing the
freelist (which means no locks at all for small strings) made things faster
again.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Fixed some anomolies in the tracing that were causing peak memory use of atom
table version to be over-reported when under load.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

richardkchapman commented:
@jakesmith @ghalliday Please review

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 85f37211ce990f164878a1c8126c5fa06b39df6a
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 92.869 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: Fail
HPCC start:
There are no components configured to run on this node ...
HPCC status:
There are no components configured to run on this node ...

Archive HPCC logs.
Removing RPM
Loaded plugins: fastestmirror
Resolving Dependencies
--> Running transaction check
---> Package hpccsystems-platform.x86_64 0:6.3.0-trunk0 will be erased
--> Finished Dependency Resolution

Dependencies Resolved

================================================================================
 Package                   Arch        Version             Repository      Size
================================================================================
Removing:
 hpccsystems-platform      x86_64      6.3.0-trunk0        installed      169 M

Transaction Summary
================================================================================
Remove  1 Package

Installed size: 169 M
Downloading packages:
Running transaction check
Running transaction test
Transaction test succeeded
Running transaction
There are no components configured to run on this node ...
\x1b[31m [ FAILED  ] 
dafilesrv is not configured to run on this node  \x1b[0m 

Removing Symlink /etc/init.d/dafilesrv                        ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/dfuplus                             ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/ecl                                 ...\x1b[32m [   OK    ]  \x1b[0m 
Removing Symlink /usr/bin/e
 ... ( comment is too long, 87 lines, 4156 bytes, truncated) ... 
ving Directory - /var/lock/HPCCSystems
Removing Directory - /var/log/HPCCSystems
Removing Directory - /var/run/HPCCSystems
Removing Directory - /var/lib/HPCCSystems
Removing user - hpcc
HPCC Uninstall: OK
Finish at: 2017-05-14_08-37-45
All done.

```

richardkchapman commented:
@AttilaVamos Looks like configgen failed to run - but not clear why (works fine on my linux and osx machines). May need access to the machine this ran on to check.

AttilaVamos commented:
@richardkchapman I will check what happened with this PR.

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Fix core on Centos because of differences in way union initializer
interpreted.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 41b101d6fa0668b879bfade65fabbf8e3e9b1019
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 95.448 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |728 |728 |0 |
|test (thor) |622 |622 |0 |
|test (roxie) |749 |749 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Add a read-only atom table for further memory reduction. This is used on
attribute names only at present, though we could consider a similar usage for
element names?

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 443b9e795e82606fd0ea1758123a6468c52eda28
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 91.664 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |728 |728 |0 |
|test (thor) |622 |622 |0 |
|test (roxie) |749 |749 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

jakesmith commented on system/jlib/jptree.ipp:
```diff
@@ -217,18 +218,168 @@ private:
+            size32_t l = strlen(key);
+            if (l <= sizeof(PTR *)-2)
+            {
+                flag=1;
```
I think worth re-emphasising 2-byte pointer alignment assumption here for clarity.

jakesmith commented on system/jlib/jptree.ipp:
```diff
@@ -217,18 +218,168 @@ private:
+#endif
+        };
+    };
+    inline PtrStrUnion<PTR>() : ptr(nullptr) {}
```
could be common constructor outside of #ifdef USE_STRUNION

richardkchapman replied:
it could though in the non-union case I think I prefer the member initializer.

richardkchapman commented on system/jlib/jptree.ipp:
```diff
@@ -217,18 +218,168 @@ private:
+    {
+        if (key)
+        {
+            size32_t l = strlen(key);
```
interesting idea, though strlen is pretty optimized where custom code to do that may not be.

richardkchapman replied:
looks like there is a strnlen function that would do the trick

ghalliday replied:
strnlen_s seems to be the standard function

richardkchapman commented on system/jlib/jptree.ipp:
```diff
@@ -366,105 +513,166 @@ protected: // data
+            unsigned table = hash % NUMTABLES;
+            CriticalBlock b(critk[table]);
+            ret = htnc[table].findh(v,hash,true);
+            if (ret->linkcount!=(unsigned short)-1)
```
It doesn't become < 0 - this is to protect against it becoming > 65535 I think

jakesmith commented on system/jlib/jptree.cpp:
```diff
@@ -73,20 +73,59 @@ IPropertyTreeIterator *createNullPTreeIterator() { return LINK(nullPTreeIterator
+#ifdef USE_READONLY_ATOMTABLE
+static const char * roAttributes[] =
+{
+#include "jptree-attrs.hpp"    // potentially auto-generated
```
should this really be an .hpp ? As it's not a strictly valid code on it's own.
Perhaps better if the whole definition was in header.

richardkchapman replied:
I didn't want to add the rest of the definition because then it's easier to generate the include via grep etc...

richardkchapman commented on system/jlib/jptree.cpp:
```diff
@@ -2954,20 +2976,28 @@ CAtomPTree::CAtomPTree(const char *_name, byte _flags, IPTArrayValue *_value, Ch
+            if (a->key.isPtr())
+                attrHT->removekey(a->key.getPtr(), nc);
+            if (a->value.isPtr())
+                attrHT->removeval(a->value.getPtr());
```
if it's not a pointer then it doesn't need destroying (is the theory)

jakesmith replied:
right, yes only has chars array member.

jakesmith commented on system/jlib/jptree.cpp:
```diff
@@ -73,20 +73,59 @@ IPropertyTreeIterator *createNullPTreeIterator() { return LINK(nullPTreeIterator
+#ifdef USE_READONLY_ATOMTABLE
+    AttrStrUnionWithTable::roNameTable = new RONameTable;
+    initializeRoTable();
+#endif
```
was there a reason why roNameTable wasn't destroyed in MODULE_EXIT ?

richardkchapman commented on system/jlib/jptree.ipp:
```diff
@@ -217,18 +218,168 @@ private:
+            if (l <= sizeof(PTR *)-2)
+            {
+                flag=1;
+                memmove(chars, key, l);  // Technically, they could overlap
```
Why would it not be? I use memmove rather than strcpy just in case someone does setAttribute(queryAttribute + 1) to strip the first char off an attr or something...

It was actually prompted by a valgrind warning that the source and dest overlapped on a strcpy (they were the same) that prompted me to think we needed to be careful.

jakesmith replied:
yes, I've seen similar flagged by valgrind.

I wasn't sure if it was safe to do  e.g. tree->setAttribute("myattr", "static string");  ?
i.e. that means : memmove(chars, "static string", l);

jakesmith commented on system/jlib/jptree.ipp:
```diff
@@ -295,9 +386,18 @@ struct PtrStrUnion
+                    int8_t idx1;
+                    int16_t idx2;
+                    int32_t idx4;
+                };
```
If 32bit , isn't this now too big?, i.e. int8+int16+in32 = 7 bytes + flag, so always 8 bytes and ptr 4.

ghalliday replied:
this would have alignment problems on big endian - fields should be in a reverse order.

jakesmith commented on system/jlib/jptree.ipp:
```diff
@@ -217,18 +218,168 @@ private:
+            char chars[sizeof(PTR *)-1];
+#else
+            char chars[sizeof(PTR *)-1];
+            char flag;
```
ignore that, silly question - 2 byte alignment bottom bit etc.

jakesmith commented:
@richardkchapman - looks good, a few questions (some of which you've already answered)

ghalliday commented on system/jlib/jptree.cpp:
```diff
@@ -3061,14 +3087,25 @@ const char *CAtomPTree::queryName() const
+    else
+    {
+        const char *name = &nameval[1];
+        size32_t nl = strlen(name);
```
you could store (len << 1) | 1 into byte[0] to avoid a few strlens.

ghalliday commented:
I've scanned it, and it looks broadly good.  Will go through it in detail later.

richardkchapman committed:
HPCC-17625 JPTree memory usage and contention issues

Minor change following review.

Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: d7526651b07b3dd613cb6cd39cf4f4404484ce3a
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 65 | 93.294 seconds |


Install hpccsystems-platform-community_6.3.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |10 |10 |0 |
|setup (thor) |10 |10 |0 |
|setup (roxie) |10 |10 |0 |
|test (hthor) |728 |728 |0 |
|test (thor) |622 |622 |0 |
|test (roxie) |749 |749 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

richardkchapman commented:
I have split this change into the memory-reduction and the contention-reduction parts. See #10007
